File: NQueen.py
class NQBacktracking:
    def __init__(self, x_, y_):
        self.ld = [0] * 30  # Left diagonal
        self.rd = [0] * 30  # Right diagonal
        self.cl = [0] * 30  # Column
        self.x = x_  # Row of the first queen
        self.y = y_  # Column of the first queen

    def printSolution(self, board):
        print("N Queen Backtracking Solution:")
        print(f"Given initial position of 1st queen at row: {self.x} column: {self.y}\n")
        for line in board:
            print(" ".join(map(str, line)))

    def solveNQUtil(self, board, col):
        if col >= N:
            return True
        if col == self.y:
            return self.solveNQUtil(board, col + 1)

        for i in range(N):
            if i == self.x:
                continue
            # Check if the queen can be placed in the current position
            if (self.ld[i - col + N - 1] != 1 and self.rd[i + col] != 1) and self.cl[i] != 1:
                board[i][col] = 1
                self.ld[i - col + N - 1] = self.rd[i + col] = self.cl[i] = 1
                if self.solveNQUtil(board, col + 1):
                    return True
                # Backtrack if placing queen doesn't lead to a solution
                board[i][col] = 0
                self.ld[i - col + N - 1] = self.rd[i + col] = self.cl[i] = 0
        return False

    def solveNQ(self):
        board = [[0 for _ in range(N)] for _ in range(N)]
        # Place the first queen at the given position
        board[self.x][self.y] = 1
        self.ld[self.x - self.y + N - 1] = self.rd[self.x + self.y] = self.cl[self.x] = 1

        if not self.solveNQUtil(board, 0):
            print("Solution does not exist")
            return False
        self.printSolution(board)
        return True


if __name__ == "__main__":
    N = 4  # Size of the chessboard (4x4)
    x, y = 1, 3  # Initial position of the first queen (row, column)
    NQBt = NQBacktracking(x, y)
    NQBt.solveNQ()

File: NQueen.cpp
#include <iostream>
#include <vector>

using namespace std;

class NQBacktracking {
public:
    vector<int> ld, rd, cl;  // Left diagonal, right diagonal, and column trackers
    int x, y;  // Position of the first queen (x - row, y - column)

    NQBacktracking(int x_, int y_) {
        ld = vector<int>(30, 0);  // Initializing diagonals and columns as 0
        rd = vector<int>(30, 0);
        cl = vector<int>(30, 0);
        x = x_;  // Row for the first queen
        y = y_;  // Column for the first queen
    }

    // Function to print the board solution
    void printSolution(vector<vector<int>>& board) {
        cout << "N Queen Backtracking Solution:" << endl;
        cout << "Given initial position of 1st queen at row: " << x
             << " column: " << y << "\n" << endl;
        for (auto& row : board) {
            for (auto& cell : row) {
                cout << cell << " ";  // Print each cell in the row
            }
            cout << endl;
        }
    }

    // Backtracking utility function to place queens
    bool solveNQUtil(vector<vector<int>>& board, int col) {
        if (col >= N) {
            return true;  // All queens are placed
        }
        if (col == y) {
            return solveNQUtil(board, col + 1);  // Skip the given column for the first queen
        }

        for (int i = 0; i < N; ++i) {
            if (i == x) {
                continue;  // Skip the row of the first queen
            }
            // Check if current position is safe for the queen
            if (ld[i - col + N - 1] != 1 && rd[i + col] != 1 && cl[i] != 1) {
                board[i][col] = 1;
                ld[i - col + N - 1] = rd[i + col] = cl[i] = 1;  // Mark diagonals and column as attacked
                if (solveNQUtil(board, col + 1)) {
                    return true;
                }
                // Backtrack if placing queen doesn't lead to a solution
                board[i][col] = 0;
                ld[i - col + N - 1] = rd[i + col] = cl[i] = 0;  // Unmark the attack
            }
        }
        return false;
    }

    // Main function to solve the N-Queens problem
    bool solveNQ() {
        vector<vector<int>> board(N, vector<int>(N, 0));  // Initialize an empty board
        board[x][y] = 1;  // Place the first queen at the given position
        ld[x - y + N - 1] = rd[x + y] = cl[x] = 1;  // Mark the diagonals and column as attacked

        if (!solveNQUtil(board, 0)) {
            cout << "Solution does not exist" << endl;
            return false;
        }
        printSolution(board);  // Print the solution
        return true;
    }

private:
    const int N = 4;  // Size of the chessboard (4x4)
};

int main() {
    int x = 1, y = 3;  // Initial position of the first queen (row 1, column 3)
    NQBacktracking NQBt(x, y);
    NQBt.solveNQ();
    return 0;
}
